### ISSUE TITLE
Import statements fail to rename app with import as PackageName.AppName

### ISSUE THREAD

**Issue by cuminandpaprika** [anz-bank/sysl#867]
### Description

Currently, sysl has syntax for importing other sysl files and altering the package and application name. However, this doesn't seem to be functioning as expected

e.g
```
import foo.sysl as Bar.Foo
import //github.com/app/repo/foo.sysl as Bar.Foo
```
This is currently failing to rename the imported packages.

### Steps to Reproduce

For this sysl file `app.sysl`
```
import //github.com/anz-bank/sysl/demo/simple/hello.sysl as ANZ.hello

App[~ignore]:
    ...

```

we run `sysl protobuf app.sysl`

**Expected behavior:**
Expect output:
`apps: {
  key: "hello"
  value: {
    name: {
      part: "hello"
    }`

**Actual behavior:**
Expect output:
`apps: {
  key: "HelloWorld"
  value: {
    name: {
      part: "HelloWorld"
    }`

### Your Environments

```sh
$ sysl info
Build:
  Version      : 0.83.0
  Git Commit   : f5b3d9a
  Date         : 2020-05-08T05:26:49Z
  Go Version   : go1.13.10 linux/amd64
  OS           : darwin/amd64
Runtime:
  GOOS/GOARCH  : darwin/amd64
  Go Version   : go1.13.10
$ sysl env
SYSL_MODULES=""
SYSL_PLANTUML="http://localhost:8080"
```

**Comment by andrewemeryanz** [comment]
I suspect the problem is more complex than a single named replacement. For example, it's possible for a single `sysl` file to contain multiple `applications` within the same file. It's also possible that an imported `sysl` file may want to rename other imported `applications`.

The background to this discussion is: How can we present a unified description of a system when an `application` may be have different names in different `sysl` files or separate `applications` in  different `sysl` files may use the same name.

For example, consider the following contrived specifications:

```
# downstream.sysl
Downstream:
    Foo:
        ...

# upstream1.sysl
import downstream.sysl as down.Down
Upstream:
    Bar:
        Down <- Foo

# upstream2.sysl
import downstream.sysl as downstream.Downstream
Upstream:
    Baz:
        Downstream <- Foo
```

Above we have two distinct `applications` that have both named themselves `Upstream`. Both applications import the same (`downstream.sysl`) file but have given them separate names. If we wanted to present a unified view of the system then we need to be able to rename imported `applications` at will.

For example, describing an `application` that consumed both `Upstream` `applications` above might want the ability to do something like the following to represent a unified system:

```
import upstream1.sysl: Upstream as up.Up1
import upstream2.sysl: Upstream as up.Up2, Downstream as down.Down
```

**Comment by cuminandpaprika** [comment]
Ah I see, thanks for the super detailed writeup! 

In the example you listed above, where `upstream1.sysl` and `upstream2.sysl` are both redeclaring the Downstream system in separate `downstream.sysl` files. Shouldn't they both referencing a single `downstream.sysl` file stored in a central repository?

**Comment by andrewemeryanz** [comment]
> Shouldn't they both referencing a single downstream.sysl file stored in a central repository?

They are both referencing the same file. They've just renamed the top-level `application` differently. Yes, in the real world `downstream.sysl` would be in a central repository. The example is used to show that the same `application` can be given different names in separate systems.

**Comment by cuminandpaprika** [comment]
Ah, that makes total sense now! Okay, that is a compelling use-case for the renaming syntax proposed above

**Comment by anzdaddy** [comment]
One option is to fix it only at the module level. Within a module, keep the same rules as today and leave it to maintainers to manage naming with that scope. Applications brought in through a module-level import would have to somehow be prefixed with that module's "name". I don't think modules have useful names at present (other than their full path) so that would need some thought.

**Comment by cuminandpaprika** [comment]
Yup, that sounds like a valid approach too. Would using package style imports like in go modules be a way to make module names more useful? 

e.g `import //github.com/anz-bank/sysl/demo/simple/hello.sysl` for local file imports?

So to summarise pros and cons:

Approach 1: Imports with extended import as syntax
- More flexibility when importing modules 
- Name conflicts must be identified and addressed by the user, rather than being built in

Approach 2: Module-level imports with name prefix
- Less manual work when naming and resolving application names.
- Could end up with very long prefixes for applications which go through several nested imports
- Seems like a lower complexity change than _Approach 1_

**Comment by andrewemeryanz** [comment]
> One option is to fix it only at the module level

What is a module in this context and how does it resolve the issue in the example above?

> Applications brought in through a module-level import

Is `import //github.com/...` a module-level import? Should the outcome be identical whether a file was imported locally or through a module?

I don't think I understand enough about how `sysl` manages `applications` internally. I understand that an `application` has a `key`, `name` and `package`. I'm not sure whether `package` and `key` are used to uniquely identify an `application` within the system or just the `key`.

**Comment by cuminandpaprika** [comment]
> I don't think I understand enough about how sysl manages applications internally. I understand that an application has a key, name and package. I'm not sure whether package and key are used to uniquely identify an application within the system or just the key.

Yeah I also have no idea what uniquely identifies a package, I think `name` is usually used to identify an application (and typically only the first part of a name, since most applications names have no NameSpaces e.g NameSpace::Appname is what creates separate AppName parts)

I've just recalled a really good concrete example for a slightly different use case:
VCards refers to CTM as Ctm, whereas Fabric refers to CTM as CTM. 

Therefore you end up with a situation where it's duplicated in the integration diagram, and shows up as 2 different services.

Only Approach 1 would resolve this issue for integration diagrams, Approach 2 would result in something like
- github.com/anzx/fabricapis/sysl/fabric.CTM
- github.com/anzx/vcards/sysl/vcards.Ctm

Both Applications would show up on integration diagrams, and both would be listed as a separate application

So to reiterate, Approach 2 solves for namespace clashes, but not for duplications.

**Comment by marcelocantos** [comment]
Packages aren't really a thing in Sysl. Everything gets shoved into a single universe. You can think of it as C `#include`. Changing that approach would require a lot of rework, and the hazards can be managed through namespaces (`a::b::c`), which is essentially how C++ solved it.

Now that we have modules, however, another option presents itself. We can tweak the import syntax to make the module structure more explicit:

```sysl
import demo = //github.com/anz-bank/sysl/demo:
    simple/hello.sysl
    world.sysl

...

MyApp:
    MyEndpoint:
        demo/hello <- greet
```

(I made up the `/` syntax off the cuff. Other suggestions are welcome.)

**Comment by cuminandpaprika** [comment]
```
import //github.com/anz-bank/sysl/demo as demo:
    simple/hello.sysl
    world.sysl

...

MyApp:
    MyEndpoint:
        demo/hello <- greet
```
I like the syntax of this, but a couple of questions I'd like to ask.

- Would this be added on top of the existing imports so that there are 2 ways of importing?

Here's my understanding of how the proposed syntax would work for a shared dependency when trying to consolidate applications for documentation generation.

Lets say we have a MobileApp and a WebApp which both share the same dependency: Server.

mobile-app.sysl
```
import //github.com/anz-bank/sysl/demo/server as SERVER:
    server.sysl

MobileApp:
   ...
```

web-app.sysl
```
import //github.com/anz-bank/sysl/demo/server as server:
    server.sysl

WebApp:
   ...
```

Now lets say we want to import both sysl files in one sysl file, in order to generate docs using `sysl-catalog`

```
import //github.com/anz-bank/sysl/demo/mobileapp:
    mobile-app.sysl
    SERVER/server.sysl
import //github.com/anz-bank/sysl/demo/webapp:
    web-app.sysl
    server/server.sysl

MyOrg:
   ...
```
How do we resolve the name conflict here, when both have been imported with different names?


We could also _drop_ the server dependency when importing, and then import it explicitly but that wouldn't repoint the CALL statements correctly e.g
```
import //github.com/anz-bank/sysl/demo/mobileapp:
    mobile-app.sysl
import //github.com/anz-bank/sysl/demo/webapp:
    web-app.sysl
import //github.com/anz-bank/sysl/demo/server:
    server.sysl

MyOrg:
   ...
```

Is it sufficient perhaps to assume the name conflict above is handled by naming conventions and best practices?

**Comment by anz-gordonj7** [comment]
Lots of discussion here, Just want to point out that the feature was added for exactly 1 use-case: Importing foreign data files (swagger/sql/etc). Without this there is no way to know how to refer to the imported types/etc.
